File: app/javascript/controllers/card_draw_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["team", "normal", "special"]

  connect() {
    this.colors = ["blue", "green", "yellow", "red", "gold", "platinum", "black"];
    this.cards = {};
    this.usedCards = {};

    // 各色のカードを初期化（仮のデータ）
    this.colors.forEach(color => {
      this.cards[color] = [`${color}-1`, `${color}-2`, `${color}-3`];
      this.usedCards[color] = [];
    });

    this.updateTeamOptions();
  }

  updateOptions() {
    this.normalTargets.forEach((normal, i) => {
      const special = this.specialTargets[i];
      const normalValue = normal.value;
      const specialValue = special.value;

      if (specialValue === "23") {
        normal.querySelectorAll("option").forEach(opt => {
          if (["11", "12", "13", "14"].includes(opt.value)) {
            opt.disabled = true;
          }
        });
      } else {
        normal.querySelectorAll("option").forEach(opt => opt.disabled = false);
      }

      if (["11", "12", "13", "14"].includes(normalValue)) {
        special.querySelectorAll("option").forEach(opt => {
          if (opt.value === "23") {
            opt.disabled = true;
            if (special.value === "23") special.value = "20";
          }
        });
      } else {
        special.querySelectorAll("option").forEach(opt => opt.disabled = false);
      }
    });
  }

  updateTeamOptions() {
    const selectedValues = this.teamTargets.map(select => select.value);

    this.teamTargets.forEach((select, i) => {
      const currentValue = select.value;
      Array.from(select.options).forEach(option => {
        option.disabled = (option.value !== currentValue && selectedValues.includes(option.value));
      });
    });
  }

  draw() {
    for (let i = 0; i < 4; i++) {
      const team = this.teamTargets[i].value;
      const normal = this.normalTargets[i].value;
      const special = this.specialTargets[i].value;

      let result = `${team} のカード結果：`;

      if (normal == 11) result += ` ブルー: ${this.drawCard("blue")} `;
      else if (normal == 12) result += ` グリーン: ${this.drawCard("green")} `;
      else if (normal == 13) result += ` イエロー: ${this.drawCard("yellow")} `;
      else if (normal == 14) result += ` レッド: ${this.drawCard("red")} `;

      if (special == 21) result += ` ゴールド: ${this.drawCard("gold")} `;
      else if (special == 22) result += ` プラチナ: ${this.drawCard("platinum")} `;
      else if (special == 23) result += ` ブラック: ${this.drawCard("black")} `;

      document.getElementById(`result${i}`).textContent = result;
    }

    document.getElementById("modal").style.display = "block";
  }

  drawCard(color) {
    if (this.cards[color].length === 0) {
      // 使用済みから再構成（リセット）
      this.cards[color] = [...this.usedCards[color]];
      this.usedCards[color] = [];
    }

    const index = Math.floor(Math.random() * this.cards[color].length);
    const card = this.cards[color][index];

    // 使用済み管理
    this.usedCards[color].push(card);
    this.cards[color].splice(index, 1);

    return card;
  }
}

File: app/controllers/card_draw_controller.rb
class CardDrawController < ApplicationController
  def index
  end

  def draw
    @results = []

    (1..4).each do |i|
      team = params["team#{i}"]
      normal_color = convert_color_key(params["normal#{i}"])
      special_color = convert_color_key(params["special#{i}"])

      cards = []

      # 通常カードの抽選
      if normal_color.present?
        card = QuestCard.where(color: normal_color, ussed: false).sample
        if card
          card.update(used: true)
          cards << "#{normal_color_jp(normal_color)}：#{card.name}"
        else
          reset_color(normal_color)
          retry_card = QuestCard.where(color: normal_color, used: false).sample
          retry_card&.update(used: true)
          cards << "#{normal_color_jp(normal_color)}：#{retry_card&.name || "カードが見つかりません"}"
        end
      end

      # スペシャルカードの抽選
      if special_color.present?
        if special_color == "black"
          cards << "ブラック：非公開カードです"
        else
          card = QuestCard.where(color: special_color, used: false).sample
          if card
            card.update(used: true)
            cards << "#{special_color_jp(special_color)}：#{card.name}"
          else
            reset_color(special_color)
            retry_card = QuestCard.where(color: special_color, used: false).sample
            retry_card&.update(used: true)
            cards << "#{special_color_jp(special_color)}：#{retry_card&.name || "カードが見つかりません"}"
          end
        end
      end

    def team_name(team_id)
      {
        "100" => "---",
        "101" => "compagno",
        "102" => "呑雲吐龍",
        "103" => "緑龍",
        "104" => "Wing Blaves",
        "105" => "Pursuing Fun!",
        "106" => "でるぽんみるちーず",
        "107" => "BITE into FATE",
        "108" => "Beiowolf",
        "109" => "SkyRoad",
        "110" => "RED ZONE!",
        "111" => "ローリエ",
        "112" => "Ne_oLion",
      }[team_id] || "不明なチーム"
    end
    @results << {
        team: team_name(team),
        cards: cards,
      }
    end

    render :result
  end

  # 補助：色名のキー変換（日本語→英語）
  def convert_color_key(jp)
    {
      "ブルー" => "blue",
      "グリーン" => "green",
      "イエロー" => "yellow",
      "レッド" => "red",
      "ゴールド" => "gold",
      "プラチナ" => "platinum",
      "ブラック" => "black",
    }[jp]
  end

  def normal_color_jp(key)
    { "blue" => "ブルー", "green" => "グリーン", "yellow" => "イエロー", "red" => "レッド" }[key]
  end

  def special_color_jp(key)
    { "gold" => "ゴールド", "platinum" => "プラチナ", "black" => "ブラック" }[key]
  end

  # 全て抽選済の色はリセット
  def reset_color(color)
    QuestCard.where(color: color).update_all(used: false)
  end
end
